engine/src/main/java/org/camunda/bpm/engine/impl/persistence/entity/TimerEntity.java

```diff
@@ -79,6 +81,14 @@ protected void preExecute(CommandContext commandContext) {
         // this timer is a repeating timer and
         // a follow up timer job has not been scheduled yet
 
+        // when reevaluateTimeCycleWhenDue is enabled and cycle is an expression
+        if (isReevaluateTimeCycleWhenDue(commandContext) && isCycleExpression()) {
+          String expressionValue = commandContext.getProcessEngineConfiguration()
+              .getExpressionManager()
+              .createExpression(jobDefinition.getJobConfiguration().substring(7))
+              .getValue(execution).toString();
+          repeat = adjustRepeatBasedOnNewExpression(expressionValue);
+        }
```

Comment:
❓ Understanding question: if the expression value does not change, we still create a new repeat, right? It simply works since the repeat doesn't change in that case, right? If a new repeat is created, how do we know to start "anew" with counting repeats, e.g. when we change from `R4PT10M` to `R3PT10M` and have already fired 3 times, how do we reset so we know to run 2 more times when we're due and recalculate the repeat from the expression?

Reply:
> Understanding question: if the expression value does not change, we still create a new repeat, right? It simply works since the repeat doesn't change in that case, right?

Correct
If a new repeat is created, we will scratch the old one and not take it into account at all (as per the feature requirements). So the timer will be fired 3 times, no matter how many times was fired before. If that's not clear from the docs, I will revisit to adjust it.

Reply:
Sorry for not being clear enough here. On a logical level (and also from the docs), that is clear to me. I was wondering, on a technical level, how we "reset" the repeat count or start date so we know that we need to start counting again for the new repeat configuration.

Reply:
After discussion, we confirmed that rework is needed for the adjusting the start time in the `repeat` field.
After the rework, the "reset" of the count is done in the `#adjustRepeatBasedOnNewExpression()` depending if the repeat interval changed, along with resetting the start date. If the repeat is the same we don't do anything. 
In case of a cron expression, we just change the replace the `repeat` since not start date is required for the ever running cron timer.